src/proxy_broker.js

```diff
@@ -0,0 +1,50 @@
+/**
+ * @license
+ * Copyright 2023 The Emscripten Authors
+ * SPDX-License-Identifier: MIT
+ */
+
+'use strict';
+
+// Map pthread IDs to message ports we use to communicate with those pthreads.
+const threadPorts = new Map();
+
+// Map recipient pthread IDs for whom we don't yet have a message port to
+// messages we've received for them.
+const bufferedMessages = new Map();
+
+function handleMessage(msg) {
+  const thread = msg.data.targetThread;
+  const port = threadPorts.get(thread);
+  if (port !== undefined) {
+    port.postMessage(msg.data, msg.data.transferList);
+  } else {
+    // Hold on to the message until we receive a port for the recipient.
+    let buffered = bufferedMessages.get(thread);
+    if (buffered === undefined) {
+      buffered = [];
+      bufferedMessages.set(thread, buffered);
+    }
+    buffered.push(msg);
```

Comment:
JS has `map.has` which is shorter for stuff like this:

```suggestion
    if (!bufferedMessages.has(thread)) {
      bufferedMessages.set(thread, []);
    }
    bufferedMessages.get(thread).push(msg);
```

But it may be slightly less efficient OTOH, I'm not sure.

Reply:
Yeah, I was trying to minimize the number of lookups to be 1 or 2 rather than 2 or 3. (I was annoyed that there's no way to do this in 1 lookup like would could in C++.) Do you think minimizing the code is more important than minimizing the lookups?

Reply:
Hmm, in general I'd guess this is not very hot code (we don't pass that many messages using postMessage). So maybe size and clarity matter more?